### PR TITLE
Extract navigator logic that waits for network idle!

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: lib/Page.js [hunk 1]
```diff
@@ -105,8 +106,11 @@ class Page extends EventEmitter {
      * @return {!Promise}
      */
     async injectFile(filePath) {
-        var expression = fs.readFileSync(filePath, 'utf8');
-        await this._client.send('Runtime.evaluate', { expression, returnByValue: true });
+        var callback;
+        var promise = new Promise(fulfill => callback = fulfill);
+        var expression = fs.readFile(filePath, 'utf8', (err, data) => callback({err, data}));
```
Comment (aslushnikov) [inline lib/Page.js]: is this change needed? The following seems to be the same: 
```js
var expression = fs.readFileSync(filePath, 'utf8');
```

File: lib/Page.js [hunk 2]
```diff
@@ -218,46 +222,32 @@ class Page extends EventEmitter {
...
      * @param {string} html
      * @return {!Promise}
      */
     async setContent(html) {
-        var resourceTree = await this._client.send('Page.getResourceTree', {});
-        await this._client.send('Page.setDocumentContent', {
-            frameId: resourceTree.frameTree.frame.id,
-            html: html
-        });
+        this.evaluate(() => document.write(html), html);
```
Comment (aslushnikov) [inline lib/Page.js]: this will work only for the first call of `setContent`, e.g.:

```js
await page.setContent('1');
await page.setContent('2');
console.log(await page.evaluate(() => document.body.innerHTML))
```

will return '12', not '2' as would be expected

File: lib/Page.js [hunk 3]
```diff
@@ -353,10 +324,17 @@ class Page extends EventEmitter {
      * @return {!Promise}
      */
     async evaluateOnInitialized(fun, ...args) {
-        var code = helpers.evaluationString(fun, args, false /* wrapInPromise */);
-        await this._client.send('Page.addScriptToEvaluateOnLoad', {
-            scriptSource: code
-        });
+        var sourceCode = Page._evaluationString(fun, ...args);
+        await this._client.send('Page.addScriptToEvaluateOnLoad', { scriptSource });
```
Comment (aslushnikov) [inline lib/Page.js]: `scriptSource` seems to be undefined here, I'd expect one of the tests to fail

**Review by aslushnikov — CHANGES_REQUESTED** [review]

File: lib/Page.js (hunk 1 above)
Comment (pavelfeldman) [reply lib/Page.js]: You are blocking all the user events (timers, etc) while reading files. Not sure if a big deal, but I try to not use any sync calls at all - they are a bad style.

File: lib/Page.js (hunk 2 above)
Comment (pavelfeldman) [reply lib/Page.js]: Good catch - I'll add document.close()

File: lib/Page.js (hunk 3 above)
Comment (pavelfeldman) [reply lib/Page.js]: It did not!

**Comment by pavelfeldman** [comment]
Comments addressed, but you should merge inline_eval first.
